File: Two_points/56.merge-intervals.go
package leetcode

import (
	"sort"
)

// @leet start
func merge(intervals [][]int) [][]int {
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})
	res := [][]int{intervals[0]}
	for _, interval := range intervals[1:] {
		last := res[len(res)-1]
		if last[1] < interval[0] {
			res = append(res, interval)
		} else {
			last[1] = max(last[1], interval[1])
		}
	}
	return res
}

// @leet end

File: TwoPointer/56.merge-intervals.py
from typing import List


# @leet start
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if len(intervals) <= 1:
            return intervals
        intervals.sort()

        def overlap(a, b):
            l, r = max(a[0], b[0]), min(a[1], b[1])
            return l <= r

        res = [intervals[0]]
        for inter in intervals[1:]:
            if overlap(inter, res[-1]):
                res[-1][1] = max(res[-1][1], inter[1])
            else:
                res.append(inter)
        return res


# @leet end

